### DISCUSSION TITLE
2024-02-23 - Source Control History - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by GordonBeeming** [SSWConsulting/SSW.Rules.Content#7993] [Chewing the Fat]
Hey SSW'ers

Let's talk about this rule!

https://ssw.com.au/rules/clean-git-history

**Comment by Freego1783** [comment]
Important topic!

Video was clear and concise.

I mainly use `git merge`, and there were cases when I squashed commits but I rarely use `git rebase`. I believe it depends on the projects and the people you're working with.

**Comment by PothieuG** [comment]
I have already used merge/squash/rebase on 3 different projects. I think the use of each depends on the guidelines and the preferences of the tech leader. Personally, I have a preference for merge; it's the one I use on a daily basis.

Otherwise, the explanations were clear in the video.

**Comment by BrookJeynes** [comment]
Video was clear. I tend to prefer squash + merge.

**Comment by ricksu978** [comment]
The video explained `git merge` vs `git rebase`.

It would be great to give a real world scenario where `git merge` or `git rebase` is preferred.

e.g.: 
Company conducts code audit every year and `git merge` is used to preserve detail history.

**Reply by AntPolkanov** [reply]
I would say `git rebase` is always preferred unless you expect lot's of conflicts.

**Comment by leahy268** [comment]
Great video for explaining how the different methods work.
Generally I prefer squash + merge.

**Comment by AshAnil** [comment]
Great video. I prefer to squash and merge commits.

**Comment by isaaclombardssw** [comment]
Video was well explained. Definitely helped to clear up any confusion I had about the differences between those 3.

**Comment by YalalCakirSSW** [comment]
Awesome rules. I've used `git merge` so far, I can see many why people preferred `squash` as well.

**Comment by synnoveisabel** [comment]
Important topic and good video explaining the different options. I usually prefer squash merge, rebase can be handy when you have 1 commit you're working on and want the latest changes in your branch, but as soon as you have more than 1 commit I find it to be a bit of a hassle.

**Comment by Calinator444** [comment]
I loved the video. I ususally get confused as to why git rebase is necessary when git merge exists, but the video made it clear. Granted it's important to observe the trade-offs when using any method in git that irreversibly modifies your commit history.

**Comment by TobyChurches** [comment]
Great Rule + Video. I mainly use `Git Merge` but Its great to actually understand `Git Rebase`

**Comment by matt-goldman** [comment]
This is hands-down the best video I've seen explaining the differences between merge strategies!

An important topic and discussion. However, it glosses over the effort it takes to "maintain a clean git history". Sometimes you rebase and have countless merge conflicts. Sometimes you have to do a soft reset just to avoid them. Git is a PITA (no denying it's better than any of the alternatives), and I posit that far more time and effort goes into to "maintaining a clean git history" on the _off chance_ that _maybe_ one day you _might_ need it, than would go into dealing with it as and when you need to.

TL;DR: squash commit ftw 👍

**Reply by drwharris** [reply]
100% agree

**Comment by sethdaily** [comment]
Informative video, and great to learn about SSW people's commit habits

**Comment by wicksipedia** [comment]
Linear commit history or GTFO!

#StrongOpinionStroglyHeld

**Comment by chrisschultzssw** [comment]
Good stuff, I didn't really know the difference between squash/merge/rebase before.
My git usage is pretty simple, but I'm sure this will be useful to me one day.

**Comment by jimmy1992abc** [comment]
All the concepts are hard for non-developers, the video is still clear and quite easy to understand.
Gordon's video in the rule is totally a different level 😶, I can feel how good it is, just too much for my year-4 knowledge.

The only project that I am involved in is the SSW Rule. Is the Squash Merge the only method?

**Comment by UlyssesMaclaren** [comment]
I finally know why i press "Squash and Merge" all the time... 

Rules and website - is there a way to make "Squash and Merge" just the default thing that happens when I approve a PR? it's a pain to keep having to press the same extra buttons every time. I never don't wanna do that.

**Comment by AlexRothwell** [comment]
In a large project, I regularly find it important that the git history is clear and thorough in its explanations. This lets me investigate the intention and reason for a change when something breaks so that any fix I make doesn't break something I didn't know about

**Comment by amankumarrr** [comment]
"Squash and Merge" is a good way to incorporate changes into the repository, as it maintains a clean history by consolidating multiple granular commits for a single functionality or fix. This approach provides the flexibility to delve deeper into the commit history if needed. Additionally, it aids in providing a high-level overview of the timeline of changes at a glance.

**Comment by vladislav-kir** [comment]
Great video and rule! I always use squash + merge, and always wondered why would people prefer rebase... Now I see where it might be used :)

**Comment by stephanfakossw** [comment]
I can see how this might be important to devs :-)

**Comment by AntPolkanov** [comment]
I have always been advocating for squashing all pull requests before merging them to the trunk 🌳 As per Linus Torvalds, "your branch - is your sandbox", so there is no need to keep that history.
If I don't expect lot's of conflicts, I `rebase` my branch on to the trunk before creating a PR.

**Comment by christoment** [comment]
Squash is the easiest to execute since it is very friendly to git history regardless of the commits practice.
Merging master to feature branch is also way easier than doing rebase, especially when trying to solve merge/rebase conflicts in a team with multiple active members. 😄

**Comment by Harry-Ross** [comment]
Great video to succinctly compare the differences between merging techniques. I usually just use `git merge`, but will give `git rebase` a go, as it seems useful to keep the commit history instead of using a merge commit.

**Comment by JackDuan123** [comment]
My reflection:

History is not that important, and it's actually expensive to keep a historical entity, we need to give our kids tech, wealth rather than history, history explains itslef in our culture or convention rather than some text record or colossom constructions.

All those text record should go to historist, they need some work to do, aren't they?

For git: JUST SQUASH PLZ!!!! keep things simple and working, stop playing fool's game

**Comment by jeoffreyfischer** [comment]
I love educational CTF like this one, it helps strengthen some fundamentals

**Comment by bahjatalaadelssw** [comment]
Great video. My git interactions are mainly Rules edits, so I just squash it, or squerge!

**Comment by ncn-ssw** [comment]
On SugarLearning project I have found the history of the Git archive invaluable in determining why decisions were made. Therefore, I am an advocate for maintaining good and easy-to-access records.

I should be easy to convince that "source tipping" is bad.

However, the topic of how to merge repositories is complex enough that simple instructions cannot be written without major warnings. If we intend to merge or split repositories using a different strategy, guidance must be provided on how to do this in a bullet-proof manner.

**Comment by bettybondoc** [comment]
Great video! I've only ever used merge cause rebase sounded scary. But now it doesn't sound so scary anymore. For SSW, my work is limited to content updates so I don't have very complex PRs. But I do agree history is important.

**Comment by danielmackay** [comment]
> 1. Do you do something else that we've missed to increase the quality of git history?

[Gitmoji](https://gitmoji.dev/)!  

I find this super valuable to be able to scan the history of a git repo and be able to visually distinguish what sort of changes have been made.

Loved the snazzy animations in the video...

**Comment by KahaMason** [comment]
Great topic.

I'm a fan of using "git merge" or "git pull" to pull latest changes from main to the branch. For PRs, squash merge is a must. Keeps things clean and simple.

**Comment by tomek-i** [comment]
I think history is important to look back at and as @ncn-ssw said figure out what and why perhaps decisions where made.
As long as it works for the Dev. Team. 

If your git history has some weird branching it might be a good indicator that something went wrong

**Comment by Dhruv-0987** [comment]
This was my favourite CTF of all time. This is a topic many people are always confused about and nobody pays attention to but its safe to say that I now know the difference between rebase and merge.

**Comment by JackDevAU** [comment]
Good detailed video! I think we currently go with Squash and merge for our internal projects but, it would be interesting to see if we can get everyone to rebase 😅😅

**Comment by jernejk** [comment]
I remember when I joined SSW and we didn't use git. :)
It took a bit of effort, but many things pushed during that time were great foundations for today, and it's great to see that we are improving.

I strongly believe that the best consistency can only be achieved with tools. AI is an amazing tool, and if we can make it so that it generates great titles and descriptions (honestly, VS one is too detailed in the wrong way, VS Code might be too abstract, but it is a good starting point).

**Comment by MarcusIrmscherSSW** [comment]
This is a bit outside my wheel house but the rule gives some helpful advise even for us amateurs. 😀

**Comment by jaydenalchin** [comment]
I subscribe to the @SamBWagner school of #Squerge

**Comment by PennyWalker** [comment]
This was a great video for someone like myself to learn about the different options. I don't use Git at this level, and it was really interesting

**Comment by MattParkerDev** [comment]
Always squerge 🏋🏼‍♂️😅🌵🧠

**Comment by NadeekaJK** [comment]
Signing off your commits is a best p[ractice to main a good commit history specially in open source projects and high security projects.
-S will sign the commit using GPG (GNU Privacy Guard) for cryptographic verification.
git commit -S -m "my message"

**Comment by Hona** [comment]
Moving from TFS to github, you can use a tool like https://github.com/git-tfs/git-tfs

& it will rewrite commits & even the authors from TFS users to GitHub users, keeping stats!
I've used this before.

**Comment by SamBWagner** [comment]
Squerge or you == 🤓.

Don't interlace git histories in repo merges though. You're losing the history of them being separated. But also, I don't think there's a correct answer. Just do whatever works.

**Comment by tombui99** [comment]
Great video! It really helps me revise the differences between Git merge, rebase and squash commits. It appears that "squash commit" is the most popular around which is almost my preferred method.

**Comment by jerryluossw** [comment]
I prefer to use `git merge`, by the way, I remember the devs had discussed it before.

**Comment by KristenHu** [comment]
I hardly use squash, it's clear but I would like to see the detailed git history, so I may still prefer git merge or git rebase.

**Comment by sylhuang** [comment]
Great video and very clearly explained with all the animation. During my work I found that git merge is most easy to use in case of conflicts. And in case of accidental push to remote, git rebase is hard to fix.

**Comment by tino-liu** [comment]
Most of the time it's git merge for me. But will look into it more in my personal project to really understand the difference between merge and rebase.

**Comment by AttackOnMorty** [comment]
My preference:
- [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)
- `git merge` to resolve conflicts (Use continuous integration to reduce conflicts)
-  `squash merge` to main

**Comment by Aibono1225** [comment]
Clear explanation in the video. Personally I prefer squerge

**Comment by drwharris** [comment]
Im a squerger!

**Comment by JeanThirion** [comment]
Squash all the way

**Comment by cvionnet** [comment]
Nice explanation. Team squerge.

**Comment by Ben0189** [comment]
That is the channel i use to learn about system design, great to have more understanding about squash merge

**Comment by jimmidier** [comment]
Squash merge is usually the preferred way to avoid any "WIP" commits

**Comment by kenzengssw** [comment]
Nice video, it definitely taught me a few things. I usually rely on git merge the most, but this knowledge will be incredibly useful.

**Comment by jakebayliss** [comment]
Usually squash commits but need to play around with rebase more. Neat little video

**Comment by awaltos** [comment]
Great vid + rule. Mainly used squash commits. Also liked to see PBI identifier (e.g. Jira number) for the task included in the squash or regular commit message to be able to easily see the context of the change where the PBI platform is different to the git platform (e.g. Jira + AWS CodeCommit 🤮)

**Comment by babakamyljanovssw** [comment]
Great video and rule that clearly explained git merge and git rebase. I've used git merge so far, and it gives full history of changes that can be valuable to track project development in most cases. Need to play with git merge and git rebase on personal projects to fully understand and see in which scenarios both of these commands suits.

**Comment by jackreimers** [comment]
Good video that covers the basics with helpful graphics.
I usually prefer to squerge
